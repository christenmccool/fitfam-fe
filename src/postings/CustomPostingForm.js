import React, {useState} from 'react';
import { useNavigate, Link as RouterLink } from 'react-router-dom';

import Box from '@mui/material/Box';
import Grid from '@mui/material/Grid';
import TextField from '@mui/material/TextField';
import Button from '@mui/material/Button';

import Alert from '../common/Alert';

/** Form for posting a new workout or editing an existing workout posting
 * 
 * date and famName supplied by the parent
 * when editing, woName and woDescription are supplied from the parent PostingEditPage 
 * 
 * PostingNewPage -> {CustomPostingForm, AddWoPostingForm}
 * PostingEditPage -> CustomPostingForm
 */
const CustomPostingForm = ({formType="create", date, famName, submitPostForm, woName, woDescription, deleteWoPost}) => {
  const navigate = useNavigate();

  const initialState = {
    woName: woName || "",
    woDescription: woDescription || ""
  }
  const [fields, setFields] = useState(initialState);
  const [errors, setErrors] = useState([]);

  const handleChange =  async (event) => {
    const {name, value} = event.target;
    setFields({...fields, [name]:value})
  }

  const handleSubmit =  async (event) => {
    event.preventDefault();

    const result = await submitPostForm(fields);

    if (result.success) {
      navigate(`/postings/${result.postId}`);
    } else {
      setErrors(result.err);
    } 
  }

  const handleDelete = () => {
    deleteWoPost();
    navigate(`/?date=${date}`);
  }

  return (
    <Box component="form" onSubmit={handleSubmit} mt={1} >
      <Grid container spacing={2} justifyContent="center">
        {formType === "edit" ?
          <>
            <Grid item xs={12} sm={6}>
              <TextField
                fullWidth
                id="date"
                name="date"
                label="Workout Post Date"
                value={date}
                disabled
              />
            </Grid>
            <Grid item xs={12} sm={6}>
              <TextField
                fullWidth
                id="famName"
                name="famName"
                label="Family name"
                value={famName}
                disabled
              />
            </Grid>
          </>
        : null
        }
        <Grid item xs={12}>
            <TextField
              fullWidth
              id="woName"
              name="woName"
              label="Workout Name (optional)"
              onChange={handleChange}
              value={fields.woName}
              InputProps={{
                inputProps: {
                  style: {fontSize: 20, textAlign: 'center', spellcheck:"false", lineHeight: "normal"}
                }
              }}
            />
        </Grid>
        <Grid item xs={12}>
          <TextField
            fullWidth
            required
            multiline
            minRows={2}
            id="woDescription"
            name="woDescription"
            label="Workout Description"
            onChange={handleChange}
            value={fields.woDescription}
            InputProps={{
              inputProps: {
                style: {fontSize: 20, textAlign: 'center', spellcheck:"false", lineHeight: "normal"}
              }
            }}
          />
        </Grid>
      </Grid>
    
      <Grid container mt={2} spacing={2} justifyContent="center">
        <Grid item>
          <Button
            type="submit"
            variant="contained"
            size="large"
          >
            Submit
          </Button>
        </Grid>
        {formType === "edit" ?
          <>
            <Grid item>
              <Button
                type="button"
                variant="outlined"
                size="large"
                onClick={handleDelete}
              >
                Delete
              </Button>
            </Grid>
            <Grid item>
              <Button
                component={RouterLink}
                to={`/?date=${date}`}
                type="button"
                size="large"
              >
                Cancel
              </Button>
            </Grid>
          </>
          : null
        }
      </Grid>

      {errors.length ?
        <Alert messages={errors} />
        : null
      } 
    </Box>
  )
}

export default CustomPostingForm;