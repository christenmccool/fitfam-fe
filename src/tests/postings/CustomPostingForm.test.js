import React from "react";
import { render, screen } from "@testing-library/react";
import UserEvent from "@testing-library/user-event";
import { MemoryRouter } from "react-router-dom";

import CustomPostingForm from "../../postings/CustomPostingForm";

it('renders without crashing', function () {
  render(
    <MemoryRouter>
      <CustomPostingForm />
    </MemoryRouter>
  );
});


it("matches snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <CustomPostingForm />
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});


it("renders posting add form", function () {
  render(
    <MemoryRouter>
      <CustomPostingForm />
    </MemoryRouter>
  );
  expect(screen.queryAllByText("Workout Name")[0]).toBeInTheDocument();  
  expect(screen.queryAllByText("Workout Description")[0]).toBeInTheDocument();  
  expect(screen.getByRole('button', {name: /submit/i})).toBeInTheDocument();
});


it("responds to user input", function () {
  render(
    <MemoryRouter>
      <CustomPostingForm />
    </MemoryRouter>
  );

  const nameField = screen.getByRole('textbox', {name: /workout name/i});
  UserEvent.type(nameField, "Custom Workout Name");
  expect(nameField.value).toEqual("Custom Workout Name");  

  const descriptionField = screen.getByRole('textbox', {name: /workout description/i});
  UserEvent.type(descriptionField, "Custom Description");
  expect(descriptionField.value).toEqual("Custom Description");  
});

