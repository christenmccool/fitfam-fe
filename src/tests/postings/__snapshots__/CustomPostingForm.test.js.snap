// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="MuiBox-root css-164r41r"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-12n0uxm-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="woName"
            id="woName-label"
          >
            Workout Name
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="woName"
              name="woName"
              style="font-size: 20px; text-align: center; line-height: normal;"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1g8cwq9"
              >
                <span>
                  Workout Name
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="woDescription"
            id="woDescription-label"
          >
            Workout Description
            <span
              aria-hidden="true"
              class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
            >
               *
            </span>
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <textarea
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
              id="woDescription"
              name="woDescription"
              required=""
              style="font-size: 20px; text-align: center; line-height: normal; height: 0px; overflow: hidden;"
            />
            <textarea
              aria-hidden="true"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
              readonly=""
              style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); font-size: 20px; text-align: center; line-height: normal; padding: 0px; width: 100%;"
              tabindex="-1"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1g8cwq9"
              >
                <span>
                  Workout Description *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-1n3sxlg-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonBase-root  css-jh47zj-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Submit
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
